Remora.Plugins/Services/PluginService.cs

```diff
@@ -1,283 +1,258 @@
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PluginService"/> class.
+    /// </summary>
+    /// <param name="options">
+    /// The service options, wrapped in an <see cref="IOptions{TOptions}"/>.
+    /// </param>
+    [Obsolete("Prefer overload which accepts PluginServiceOptions directly.")]
+    public PluginService(IOptions<PluginServiceOptions> options)
+        : this(options.Value)
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PluginService"/> class.
+    /// </summary>
+    /// <param name="options">
+    /// The service options.
+    /// </param>
+    public PluginService(PluginServiceOptions options)
+    {
+        _options = options;
+        _pluginLoader = new PluginLoader();
+        _pluginDescriptors = new Dictionary<string, IPluginDescriptor>();
+        _fileSystemWatcher = new FileSystemWatcher(GetApplicationDirectory(), _options.Filter)
+        {
+            EnableRaisingEvents = true,
+            IncludeSubdirectories = true,
+            NotifyFilter = NotifyFilters.Attributes
+                           | NotifyFilters.CreationTime
+                           | NotifyFilters.DirectoryName
+                           | NotifyFilters.FileName
+                           | NotifyFilters.LastAccess
+                           | NotifyFilters.LastWrite
+                           | NotifyFilters.Security
+                           | NotifyFilters.Size,
+        };
+        _fileSystemWatcher.Changed += (_, e) =>
+        {
+            // Unload old plugin (file changed).
+            UnloadPlugin(Path.GetFileNameWithoutExtension(e.FullPath));
+
+            // Load new one.
+            LoadPlugin(e.FullPath, out var result);
+            if (!result.IsSuccess)
+            {
+                _options.ErrorDelegate?.Invoke(result);
+            }
+        };
+        _fileSystemWatcher.Created += (_, e) =>
+        {
+            // Load Plugin (file created).
+            LoadPlugin(e.FullPath, out var result);
+            if (!result.IsSuccess)
+            {
+                _options.ErrorDelegate?.Invoke(result);
+            }
+        };
+        _fileSystemWatcher.Deleted += (_, e) =>
+        {
+            // Unload plugin (file deleted).
+            UnloadPlugin(Path.GetFileNameWithoutExtension(e.FullPath));
+        };
+        _fileSystemWatcher.Renamed += (_, e) =>
+        {
+            // Unload old plugin (file renamed).
+            UnloadPlugin(Path.GetFileNameWithoutExtension(e.OldFullPath));
+
+            // Load new one.
+            LoadPlugin(e.FullPath, out var result);
+            if (!result.IsSuccess)
+            {
+                _options.ErrorDelegate?.Invoke(result);
+            }
+        };
+    }
+
+    /// <summary>
+    /// Loads all available plugins with a specific filter and watches them for any changes.
+    /// </summary>
+    public void LoadPlugins()

```

Comment:
Somewhere in my bot it crashes inside of here. I might have overlooked something here.